src/main/java/net/dv8tion/jda/bot/sharding/DefaultShardManager.java

```diff
@@ -0,0 +1,587 @@
+        this.maxReconnectDelay = maxReconnectDelay;
+        this.corePoolSize = corePoolSize;
+        this.enableVoice = enableVoice;
+        this.shutdownHook = enableShutdownHook ? new Thread(this::shutdown, "JDA Shutdown Hook") : null;
+        this.enableBulkDeleteSplitting = enableBulkDeleteSplitting;
+        this.autoReconnect = autoReconnect;
+        this.idleProvider = idleProvider;
+        this.retryOnTimeout = retryOnTimeout;
+
+        if (shardsTotal != -1)
+        {
+            if (shardIds == null)
+            {
+                this.shards = new ShardCacheViewImpl(shardsTotal);
+                for (int i = 0; i < this.shardsTotal; i++)
+                    this.queue.add(i);
+            }
+            else
+            {
+                this.shards = new ShardCacheViewImpl(shardIds.size());
+                shardIds.stream().distinct().sorted().forEach(this.queue::add);
+            }
+        }
+    }
+
+    @Override
+    public void addEventListener(final Object... listeners)
+    {
+        ShardManager.super.addEventListener(listeners);
+        Collections.addAll(this.listeners, listeners);
+    }
+
+    @Override
+    public void removeEventListener(final Object... listeners)
+    {
+        ShardManager.super.removeEventListener(listeners);
+        this.listeners.removeAll(Arrays.asList(listeners));
+    }
+
+    @Override
+    public int getShardsQueued()
+    {
+        return this.queue.size();
+    }
+
+    @Override
+    public ShardCacheView getShardCache()
+    {
+        return this.shards;
+    }
+
+    public void login() throws LoginException, IllegalArgumentException
+    {
+        // building the first one in the current thread ensures that LoginException and IllegalArgumentException can be thrown on login
+        JDAImpl jda = null;
+        try
+        {
+            final int shardId = this.queue.isEmpty() ? 0 : this.queue.peek();
+
+            jda = this.buildInstance(shardId);
+            this.shards.getMap().put(shardId, jda);
+            this.queue.remove(shardId);
+        }
+        catch (final RateLimitedException e)
+        {
+            // add not remove 'shardId' from the queue and try the first one again after 5 seconds in the async thread
+        }
+        catch (final Exception e)
+        {
+            if (jda != null)
+                jda.shutdown();
+            throw e;
+        }
+
+        this.worker = this.executor.scheduleWithFixedDelay(this::processQueue, 5000, 5000, TimeUnit.MILLISECONDS); // 5s for ratelimit
+
+        if (this.shutdownHook != null)
+            Runtime.getRuntime().addShutdownHook(this.shutdownHook);
+    }
+
+    @Override
+    public void restart(final int shardId)
+    {
+        Checks.notNegative(shardId, "shardId");
+        Checks.check(shardId < this.shardsTotal, "shardId must be lower than shardsTotal");
+
+        final JDA jda = this.shards.getMap().remove(shardId);
+        if (jda != null)
+            jda.shutdown();
```

Comment:
You might want to either use shutdownNow() or give the user the option to use shutdownNow() as in large bots this can make quite the difference on the time it takes from rebooting a shard for it to be unavaliable (example when on an error the shard starts spamming "unknown..."